File: frontend/src/services/genetica/Reproductores.ts
// src/services/genetica/Reproductores.ts
import API_BASE from "../../config/api";

export type Verraco = {
  id: number;
  identificacion: string;
  raza: string;
  fechaNacimiento: string;
  origen: string;
  padre: string;
  madre: string;
  peso: number;
  estadoReproductivo: "Activo" | "Reposo" | "Baja";
  salud: string;
  valorGenetico: string;
  observaciones: string;
};

export type NuevoVerraco = Omit<Verraco, "id">;

const EMPRESA_ID = 1;
const GRANJA_ID = 1;

function mapApiToVerraco(api: any): Verraco {
  return {
    id: api.id,
    identificacion: api.identificacion,
    raza: api.raza,
    fechaNacimiento: api.fechaNacimiento,
    origen: api.origen,
    padre: api.padre,
    madre: api.madre,
    peso: api.peso,
    estadoReproductivo: api.estadoReproductivo,
    salud: api.salud,
    valorGenetico: api.valorGenetico,
    observaciones: api.observaciones ?? "",
  };
}

export async function getVerracos(): Promise<Verraco[]> {
  const res = await fetch(
    `${API_BASE}/genetica/verracos/?empresa_id=${EMPRESA_ID}&granja_id=${GRANJA_ID}`
  );
  if (!res.ok) {
    throw new Error("Error al obtener verracos");
  }
  const data = await res.json();
  return data.map(mapApiToVerraco);
}

export async function addVerraco(verraco: NuevoVerraco): Promise<Verraco> {
  const payload = {
    empresa_id: EMPRESA_ID,
    granja_id: GRANJA_ID,
    ...verraco,
  };

  const res = await fetch(`${API_BASE}/genetica/verracos/`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    throw new Error("Error al registrar verraco");
  }
  const data = await res.json();
  return mapApiToVerraco(data);
}

export async function updateVerraco(
  id: number,
  verraco: NuevoVerraco
): Promise<Verraco> {
  const payload = { ...verraco };

  const res = await fetch(
    `${API_BASE}/genetica/verracos/${id}?empresa_id=${EMPRESA_ID}&granja_id=${GRANJA_ID}`,
    {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    }
  );
  if (!res.ok) {
    throw new Error("Error al actualizar verraco");
  }
  const data = await res.json();
  return mapApiToVerraco(data);
}

export async function deleteVerraco(id: number): Promise<void> {
  const res = await fetch(
    `${API_BASE}/genetica/verracos/${id}?empresa_id=${EMPRESA_ID}&granja_id=${GRANJA_ID}`,
    { method: "DELETE" }
  );
  if (!res.ok) {
    throw new Error("Error al eliminar verraco");
  }
}

File: backend/services/genetica/Reproductores.py
// src/services/geneticaReproductoresService.ts
export type Verraco = {
  id: string;
  identificacion: string;
  raza: string;
  fechaNacimiento: string;
  origen: string;
  padre: string;
  madre: string;
  peso: number;
  estadoReproductivo: "Activo" | "Reposo" | "Baja";
  salud: string;
  valorGenetico: string;
  observaciones: string;
};

const STORAGE_KEY = "genetica_reproductores";

function getAll(): Verraco[] {
  const raw = localStorage.getItem(STORAGE_KEY);
  return raw ? JSON.parse(raw) : [];
}

function saveAll(lista: Verraco[]) {
  localStorage.setItem(STORAGE_KEY, JSON.stringify(lista));
}

export function getVerracos() {
  return getAll();
}

export function addVerraco(verraco: Omit<Verraco, "id">) {
  const nuevo: Verraco = { ...verraco, id: crypto.randomUUID() };
  const verracos = getAll();
  verracos.push(nuevo);
  saveAll(verracos);
  return nuevo;
}

export function updateVerraco(id: string, verraco: Omit<Verraco, "id">) {
  const verracos = getAll();
  const idx = verracos.findIndex(v => v.id === id);
  if (idx >= 0) {
    verracos[idx] = { ...verraco, id };
    saveAll(verracos);
    return verracos[idx];
  }
  throw new Error("Registro no encontrado");
}

export function deleteVerraco(id: string) {
  const verracos = getAll().filter(v => v.id !== id);
  saveAll(verracos);
}
